File: admin-dashboard/src/services/apiClient.js
const API_BASE_URL = (import.meta.env.VITE_API_BASE_URL || 'http://127.0.0.1:8000').replace(/\/$/, '')
const TOKEN_KEY = 'safezone.admin.access_token'
const PROFILE_KEY = 'safezone.admin.profile'

export class ApiError extends Error {
  constructor(message, status, details = null) { super(message); this.name = 'ApiError'; this.status = status; this.details = details }
}
export const hasToken = () => Boolean(sessionStorage.getItem(TOKEN_KEY))
export const getStoredProfile = () => { try { return JSON.parse(sessionStorage.getItem(PROFILE_KEY)) } catch { return null } }
export const storeProfile = (profile) => sessionStorage.setItem(PROFILE_KEY, JSON.stringify(profile))
export const clearSession = () => { sessionStorage.removeItem(TOKEN_KEY); sessionStorage.removeItem(PROFILE_KEY) }

function errorMessage(payload, fallback) {
  if (typeof payload?.detail === 'string') return payload.detail
  if (Array.isArray(payload?.detail)) return payload.detail.map((item) => item.msg).join('; ')
  return fallback
}
async function request(path, options = {}) {
  const headers = new Headers(options.headers || {})
  const token = sessionStorage.getItem(TOKEN_KEY)
  if (token) headers.set('Authorization', `Bearer ${token}`)
  if (options.body && !(options.body instanceof FormData) && !headers.has('Content-Type')) headers.set('Content-Type', 'application/json')
  let response
  try { response = await fetch(`${API_BASE_URL}${path}`, { ...options, headers }) }
  catch { throw new ApiError('Cannot reach the SafeZone API. Confirm the backend is running.', 0) }
  const payload = response.status === 204 ? null : await response.json().catch(() => null)
  if (!response.ok) {
    if (response.status === 401 && token) window.dispatchEvent(new Event('safezone:auth-expired'))
    throw new ApiError(errorMessage(payload, `Request failed (${response.status})`), response.status, payload)
  }
  return payload
}
function query(params) {
  const values = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => { if (value !== '' && value !== null && value !== undefined) values.set(key, value) })
  return values.size ? `?${values}` : ''
}

export const api = {
  async login(email, password) {
    const body = new URLSearchParams({ username: email, password })
    const token = await request('/auth/login', { method: 'POST', body, headers: { 'Content-Type': 'application/x-www-form-urlencoded' } })
    sessionStorage.setItem(TOKEN_KEY, token.access_token)
    try {
      const profile = await request('/auth/me')
      if (profile.role !== 'ADMIN') throw new ApiError('This account is not an administrator.', 403)
      return profile
    } catch (error) { clearSession(); throw error }
  },
  me: () => request('/auth/me'),
  crimes: (filters = {}) => request(`/crimes${query(filters)}`),
  createCrime: (data) => request('/crimes', { method: 'POST', body: JSON.stringify(data) }),
  updateCrime: (id, data) => request(`/crimes/${id}`, { method: 'PATCH', body: JSON.stringify(data) }),
  deactivateCrime: (id) => request(`/crimes/${id}`, { method: 'DELETE' }),
  importCrimes: (file) => { const body = new FormData(); body.append('file', file); return request('/crimes/import/csv', { method: 'POST', body }) },
  activePrps: () => request('/prp/active'),
  generatePrps: (data) => request('/prp/generate', { method: 'POST', body: JSON.stringify(data) }),
  approvePrps: (id) => request(`/prp/runs/${id}/approve`, { method: 'POST' }),
  activatePrps: (id) => request(`/prp/runs/${id}/activate`, { method: 'POST' }),
  assignments: (params = {}) => request(`/patrols/assignments${query(params)}`),
  automaticAssignments: (id) => request('/patrols/assignments/automatic', { method: 'POST', body: JSON.stringify({ optimization_run_id: id }) }),
  overrideAssignment: (id, data) => request(`/patrols/assignments/${id}/override`, { method: 'PATCH', body: JSON.stringify(data) }),
  cancelAssignment: (id) => request(`/patrols/assignments/${id}/cancel`, { method: 'POST' }),
}
export { API_BASE_URL }

File: mobile/safezone_app/lib/services/api_client.dart
import 'dart:convert';

import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:http/http.dart' as http;
import 'package:safezone_app/models/patrol_assignment.dart';
import 'package:safezone_app/models/user_profile.dart';

abstract interface class TokenStore {
  Future<String?> read();
  Future<void> write(String token);
  Future<void> clear();
}

class SecureTokenStore implements TokenStore {
  const SecureTokenStore();
  static const _storage = FlutterSecureStorage(aOptions: AndroidOptions());
  static const _key = 'safezone_police_access_token';

  @override
  Future<String?> read() => _storage.read(key: _key);
  @override
  Future<void> write(String token) => _storage.write(key: _key, value: token);
  @override
  Future<void> clear() => _storage.delete(key: _key);
}

class ApiException implements Exception {
  const ApiException(this.message, {this.statusCode});
  final String message;
  final int? statusCode;
  @override
  String toString() => message;
}

class ApiClient {
  ApiClient({
    http.Client? client,
    TokenStore? tokenStore,
    String? baseUrl,
  })  : _client = client ?? http.Client(),
        _tokens = tokenStore ?? const SecureTokenStore(),
        baseUrl = (baseUrl ?? const String.fromEnvironment(
          'SAFEZONE_API_BASE_URL',
          defaultValue: 'http://10.0.2.2:8000',
        )).replaceAll(RegExp(r'/$'), '');

  final http.Client _client;
  final TokenStore _tokens;
  final String baseUrl;

  Future<UserProfile> login(String email, String password) async {
    final response = await _client.post(
      Uri.parse('$baseUrl/auth/login'),
      headers: const {'Content-Type': 'application/x-www-form-urlencoded'},
      body: {'username': email.trim(), 'password': password},
    );
    final payload = _decode(response);
    if (response.statusCode != 200) {
      throw ApiException(_message(payload, 'Login failed'), statusCode: response.statusCode);
    }
    await _tokens.write(payload['access_token'] as String);
    try {
      final profile = await currentUser();
      if (profile.role != 'POLICE') {
        throw const ApiException('This account is not authorized for the Police application.');
      }
      return profile;
    } catch (_) {
      await _tokens.clear();
      rethrow;
    }
  }

  Future<UserProfile> currentUser() async =>
      UserProfile.fromJson(await _authorizedJson('GET', '/auth/me'));

  Future<PatrolAssignment?> currentAssignment() async {
    try {
      return PatrolAssignment.fromJson(
        await _authorizedJson('GET', '/patrols/assignments/current'),
      );
    } on ApiException catch (error) {
      if (error.statusCode == 404) return null;
      rethrow;
    }
  }

  Future<PatrolAssignment> transition(String assignmentId, String action) async =>
      PatrolAssignment.fromJson(
        await _authorizedJson('POST', '/patrols/assignments/$assignmentId/$action'),
      );

  Future<void> logout() => _tokens.clear();

  Future<Map<String, dynamic>> _authorizedJson(String method, String path) async {
    final token = await _tokens.read();
    if (token == null) throw const ApiException('Your session has expired.', statusCode: 401);
    final uri = Uri.parse('$baseUrl$path');
    final headers = {'Authorization': 'Bearer $token', 'Accept': 'application/json'};
    final response = method == 'POST'
        ? await _client.post(uri, headers: headers)
        : await _client.get(uri, headers: headers);
    final payload = _decode(response);
    if (response.statusCode < 200 || response.statusCode >= 300) {
      if (response.statusCode == 401) await _tokens.clear();
      throw ApiException(_message(payload, 'SafeZone request failed'), statusCode: response.statusCode);
    }
    return payload;
  }

  Map<String, dynamic> _decode(http.Response response) {
    if (response.body.isEmpty) return <String, dynamic>{};
    try {
      return jsonDecode(response.body) as Map<String, dynamic>;
    } on FormatException {
      throw ApiException('The SafeZone API returned an unreadable response.', statusCode: response.statusCode);
    }
  }

  String _message(Map<String, dynamic> payload, String fallback) {
    final detail = payload['detail'];
    if (detail is String) return detail;
    if (detail is List) {
      return detail.map((item) => item is Map ? item['msg'] : item).join('; ');
    }
    return fallback;
  }
}
